
# Parallel gene set signature reversion reveals drug repurposing candidates for human cancers

Here we compare and contrast three methods for identifying disease signatures for downstream signature reversion drug repurposing candidate identification and demonstrate their ability to yield many novel drug repurposing candidates for the cancers with the lowest relative survival rates. Because our methodology relies on patient tumor RNA-Seq profiles from The Cancer Genome Atlas (TCGA) [1–4], LINCS 2020 data [5], and cancer-specific cell line omic profiles from DepMAP[6] and PRISM[6] we focused on low survival cancers where this data was available, namely pancreatic adenocarcinoma (5-year survival of less than 8% [7] liver hepatocellular carcinoma (5-year survival of 18%)[8], glioblastoma (5-year survival rate of 4%) [9], and lung adenocarcinoma (5-year survival of 5% for metastatic cases) [10]. By evaluating the different methods across these cancers, we found that the various methods detected distinct disease signatures to determine candidates, resulting in mostly unique drug candidates. We conclude that one disease signature method did not outperform the others. Future studies should consider using all these methods to determine drug repurposing candidates or develop a strategy to combine these disease signatures.


## Authors

- [Jennifer L. Fisher](https://www.github.com/JenFisher7), [Elizabeth J. Wilk](), [Victoria L Flanary](), [Vishal H. Oza](), [Brittany N. Lasseigne]()

Department of Cell, Developmental and Integrative Biology, Heersink School of Medicine, University of Alabama at Birmingham, Birmingham, AL, 35294, USA.



## Roadmap

- Additional browser support

- Add more integrations


## Lasseigne Lab 
![Logo](https://www.lasseigne.org/img/main/lablogo.png)


## Acknowledgements

 - [Awesome Readme Templates](https://awesomeopensource.com/project/elangosundar/awesome-README-templates)
 - [Awesome README](https://github.com/matiassingers/awesome-readme)
 - [How to write a Good readme](https://bulldogjob.com/news/449-how-to-write-a-good-readme-for-your-github-project)


## References

1. Cancer Genome Atlas Research Network. Electronic address: andrew_aguirre@dfci.harvard.edu, Cancer Genome Atlas Research Network. Integrated Genomic Characterization of Pancreatic Ductal Adenocarcinoma. Cancer Cell. 2017;32:185–203.e13.
2. Cancer Genome Atlas Research Network. Electronic address: wheeler@bcm.edu, Cancer Genome Atlas Research Network. Comprehensive and Integrative Genomic Characterization of Hepatocellular Carcinoma. Cell. 2017;169:1327–41.e23.
3. Cancer Genome Atlas Research Network. Comprehensive molecular profiling of lung adenocarcinoma. Nature. 2014;511:543–50.
4. Brennan CW, Verhaak RGW, McKenna A, Campos B, Noushmehr H, Salama SR, et al. The somatic genomic landscape of glioblastoma. Cell. 2013;155:462–77.
5. Yang C, Zhang H, Chen M, Wang S, Qian R, Zhang L, et al. A survey of optimal strategy for signature-based drug repositioning and an application to liver cancer. Elife [Internet]. 2022;11. Available from: http://dx.doi.org/10.7554/eLife.71880
6. DepMap: The Cancer Dependency Map Project at Broad Institute [Internet]. [cited 2022 May 25]. Available from: https://depmap.org/portal/depmap/
7. Orth M, Metzger P, Gerum S, Mayerle J, Schneider G, Belka C, et al. Pancreatic ductal adenocarcinoma: biological hallmarks, current status, and future perspectives of combined modality treatment approaches. Radiat Oncol. 2019;14:141.
8. Villanueva A. Hepatocellular Carcinoma. N Engl J Med. 2019;380:1450–62.
9. Poon MTC, Sudlow CLM, Figueroa JD, Brennan PM. Longer-term (≥ 2 years) survival in patients with glioblastoma in population-based studies pre- and post-2005: a systematic review and meta-analysis. Sci Rep. 2020;10:11622.
10. Huang C-Y, Chen B-H, Chou W-C, Yang C-T, Chang JW-C. Factors associated with the prognosis and long-term survival of patients with metastatic lung adenocarcinoma: a retrospective analysis. J Thorac Dis. 2018;10:2070–8.



## License

[MIT](https://choosealicense.com/licenses/mit/)

