---
title: "211213_recount3_download"
author: "Jennifer Fisher"
date: "12/13/2021"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# START 
    Goal/Purpose of operations: 
    Donwloading Recount 3 RNA-seq Counts and normlizing counts by gene length via tpm 
    
    Finished psedocode on: Date
    211213
    
    System which operations were done on:
    my laptop(on cheaha)
    
    GitHub Repo:
    Transfer_Learning_R03
    
    Docker/Singularity: 
    rstudio_tf_dr_v3
    
    Directory of operations: 
    /data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03
    
    Scripts being edited for operations:
    NA
    
    Data being used:
    recount3
    recount3 is an online resource consisting of RNA-seq gene, exon, and exon-exon junction counts as well as coverage bigWig files for 8,679 and 10,088 different studies for human and mouse respectively. It is the third generation of the ReCount project and part of recount.bio.
    http://rna.recount.bio/
    
    Papers and tools: 
    recount3- see details above 

# STEPS
### Set working directory 
/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03

### load in data
in analysis

### Analysis
```{r}
library(recount3)
```

```{r eval= FALSE}
human_projects<- available_projects()
```

#GBM

```{r eval= FALSE}
 human_recount_projects_gbm <- human_projects[! (human_projects$project == "GBM" | human_projects$project == "BRAIN"),]
```

```{r eval= FALSE}
test<- create_rse(human_recount_projects_gbm[1,])
recount3__gbm_tf <- as.data.frame(assay(test))
```


```{r eval= FALSE}
human_projects$gene_url <- apply(human_projects, 1, function(x)
    locate_url(
        project = x["project"],
        project_home = x["project_home"],
        type = "gene",
        organism = x["organism"],
        annotation = annotation_options(x["organism"])[1] # Use default annotation
    ))
```

```{r eval= FALSE}
human_projects$command <- paste0("wget ", human_projects$gene_url, sep="")
```


```{r eval= FALSE}
saveRDS(human_projects, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/human_samples.rds")
```

```{r eval= FALSE}
write.table(human_projects$command[1:1000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_1.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[1001:2000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_2.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[2001:3000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_3.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[3001:4000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_4.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[4001:5000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_5.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[5001:6000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_6.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[6001:7000], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_7.sh", row.names= FALSE, quote= FALSE)
write.table(human_projects$command[7001:8742], file = "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gbm_gene_url_8.sh", row.names= FALSE, quote= FALSE)

```


```{r eval= FALSE}
count_table<- function(file_dir){
  setwd(file_dir)
  files <- list.files(file_dir)
  
  test<- read.table( files[1], header= TRUE)
   for (i in 2:length(files)){
     test2<- read.table( files[i], header= TRUE)
      #print(colnames(test2))
     #print(test2)
     #colnames(test2)
    # print(colnames(test2))
    #print(colnames(test))
     test<- cbind(test, test2)
     #print(files[i])
   }
  test<- test[,!grepl("gene_id", colnames(test))]
  return(test)
}
```


```{r eval= FALSE}
group1<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_1/")
```

```{r}
saveRDS(group1, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group1.rds")
rm(group1)
gc()
```



```{r eval= FALSE}
group2<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_2/")
saveRDS(group2, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group2.rds")
rm(group2)
gc()
group3<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_3/")
saveRDS(group3, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group3.rds")
rm(group3)
gc()
group4<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_4/")
saveRDS(group4, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group4.rds")
rm(group4)
gc()
group5<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_5/")
saveRDS(group5, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group5.rds")
rm(group5)
gc()
group6<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_6/")
saveRDS(group6, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group6.rds")
rm(group6)
gc()
group7<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_7/")
saveRDS(group7, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group7.rds")
rm(group7)
gc()
group8<- count_table("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/gene_url_8/")
saveRDS(group8, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group8.rds")
rm(group8)
gc()
```



```{r eval= FALSE}
recount3_group1 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group1.rds")
recount3_group2 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group2.rds")
recount3_group3 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group3.rds")
recount3_group4 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group4.rds")
recount3_group5 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group5.rds")
recount3_group6 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group6.rds")
recount3_group7 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group7.rds")
recount3_group8 <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group8.rds")
```

```{r eval= FALSE}
#recount3_Counts<- cbind(recount3_group1, recount3_group2, recount3_group3, recount3_group4, recount3_group5, recount3_group6, recount3_group7, recount3_group8)
#rm(recount3_group1, recount3_group2, recount3_group3, recount3_group4, recount3_group5, recount3_group6, recount3_group7, recount3_group8)
#saveRDS(recount3_Counts, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_counts.rds")
#crashes R
#gc()
```

#change it to tpm 

```{r eval= FALSE}
Counts_to_tpm <- function(counts, featureLength) {
  # Ensure valid arguments.
  stopifnot(length(featureLength) == nrow(counts))
  # Compute effective lengths of features in each library.
  effLen <- featureLength
  # Process one column at a time.
  tpm <- do.call(cbind, lapply(1:ncol(counts), function(i) {
    rate = log(counts[,i]) - log(effLen)
    denom = log(sum(exp(rate)))
    exp(rate - denom + log(1e6))
  }))
  # Copy the row and column names from the original matrix.
  colnames(tpm) <- colnames(counts)
  rownames(tpm) <- rownames(counts)
  return(tpm)
}
```



```{r eval= FALSE}
#SRP118922
test<- create_rse(human_projects[(human_projects$project == "SRP118922"),])
```
```{r eval= FALSE}
gene_info <- test@rowRanges
rm(test)
```

```{r eval= FALSE}
recount3_group1_tpm<- Counts_to_tpm(recount3_group1[,-1], gene_info$bp_length)
rownames(recount3_group1_tpm)<- recount3_group1[,1]
saveRDS(recount3_group1_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group1_tpm.rds")
rm(recount3_group1)
```

```{r eval= FALSE}
recount3_group2_tpm<- Counts_to_tpm(as.data.frame(recount3_group2[,-1]) , gene_info$bp_length)
rownames(recount3_group2_tpm)<- recount3_group2[,1]
saveRDS(recount3_group2_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group2_tpm.rds")
rm(recount3_group2)
rm(recount3_group2_tpm)
```
```{r eval= FALSE}
recount3_group3_tpm<- Counts_to_tpm(recount3_group3[,-1] , gene_info$bp_length)
rownames(recount3_group3_tpm)<- recount3_group3[,1]
saveRDS(recount3_group3_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group3_tpm.rds")
rm(recount3_group3)
rm(recount3_group3_tpm)
```

```{r eval= FALSE}
recount3_group4_tpm<- Counts_to_tpm(recount3_group4[,-1] , gene_info$bp_length)
rownames(recount3_group4_tpm)<- recount3_group4[,1]
saveRDS(recount3_group4_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group4_tpm.rds")
rm(recount3_group4)
rm(recount3_group4_tpm)
```

```{r eval= FALSE}
recount3_group5_tpm<- Counts_to_tpm(recount3_group5[,-1] , gene_info$bp_length)
rownames(recount3_group5_tpm)<- recount3_group5[,1]
saveRDS(recount3_group5_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group5_tpm.rds")
rm(recount3_group5)
rm(recount3_group5_tpm)
```
```{r eval= FALSE}
recount3_group6_tpm<- Counts_to_tpm(recount3_group6[,-1] , gene_info$bp_length)
rownames(recount3_group6_tpm)<- recount3_group6[,1]
saveRDS(recount3_group6_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group6_tpm.rds")
rm(recount3_group6)
rm(recount3_group6_tpm)
```
```{r eval= FALSE}
recount3_group7_tpm<- Counts_to_tpm(recount3_group7[,-1] , gene_info$bp_length)
rownames(recount3_group7_tpm)<- recount3_group7[,1]
saveRDS(recount3_group7_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group7_tpm.rds")
rm(recount3_group7)
rm(recount3_group7_tpm)
```

```{r eval= FALSE}
recount3_group8_tpm<- Counts_to_tpm(recount3_group8[,-1] , gene_info$bp_length)
rownames(recount3_group8_tpm)<- recount3_group8[,1]
saveRDS(recount3_group8_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group8_tpm.rds")
rm(recount3_group8)
rm(recount3_group8_tpm)
```

there was no reason for the removal of the first col form the matrix add ing it back as group 9 done 220728 
```{r}
#"DRR002593" "DRR001796" "DRR000897" "DRR031682" "DRR006400" "DRR001622" "DRR001173"
recount3_group9 <- cbind(recount3_group1[,1], recount3_group2[,1], recount3_group3[,1], recount3_group4[,1], recount3_group5[,1], recount3_group6[,1], recount3_group7[,1], recount3_group8[,1])
recount3_group9_tpm <- Counts_to_tpm(recount3_group9 , gene_info$bp_length)
colnames(recount3_group9_tpm)<- c( colnames(recount3_group1)[1], colnames(recount3_group2)[1], colnames(recount3_group3)[1], colnames(recount3_group4)[1], colnames(recount3_group5)[1], colnames(recount3_group6)[1], colnames(recount3_group7)[1], colnames(recount3_group8)[1])
saveRDS(recount3_group9_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group9_tpm.rds")
```


### Save Data
Done in analysis 

    
# END
    Location of final scripts:
    ./scripts
    
    Location of data produced:
    ./data/recount3
    
    Dates when operations were done:
    211213-220104 
    
## Versions
```{r}
sessionInfo()
```

