---
title: "220112_gbm_deseq2_transfer_learning"
author: "Jennifer Fisher"
date: "1/12/2022"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# START 
    Goal/Purpose of operations: 
    I am taking the latent variables from PLIER for GBM and applying it to the GBM tumor and controls. Also using GTEX controls for this experiment, so I am not only looking at 5 controls. Note DESeq2 was also applied in this script
    
    Finished psedocode on: 
    220112
    
    System which operations were done on: 
    my laptop
    
    GitHub Repo:
    Transfer_Learning_R03
    
    Docker/Singularity: 
    rstudio_cancer_dr
    
    Directory of operations:
    "/home/rstudio"
    
    Scripts being edited for operations:
    using some modified scripts from MultiPLIER paper
    
    Data being used: 
    TCGA GBM data and GTEx Brain samples 
    PLIER output 
    
    Papers and tools:
    limma and MultiPLIER
    PCA analysis
    DESeq2

# STEPS

### load in data
```{r}
library(recount3)
library(SummarizedExperiment)
library(dplyr)
library(stringr)
```
load in function scripts 
```{r}
source("/home/rstudio/script/functions_cancer_signature_reversion_JLF.R")
source("/home/rstudio/script/plier_util.R")
source("/home/rstudio/script/test_LV_differences.R")
```

```{r}
gbm_tcga_counts <- readRDS("~/data/recount3/recount3_fix_download/gbm_tcga_counts.rds")
brain_gtex_counts <- readRDS("~/data/recount3/recount3_fix_download/brain_gtex_counts.rds")
feature_info <- readRDS("~/data/recount3/recount3_fix_download/feature_info.rds")
gbm_tcga_metadata <- readRDS("~/data/recount3/recount3_fix_download/gbm_tcga_metadata.rds")
brain_gtex_metadata <- readRDS("~/data/recount3/recount3_fix_download/brain_gtex_metadata.rds")
```

check counts and metadata info
```{r}
dim(gbm_tcga_counts)
dim(gbm_tcga_metadata)
```

```{r}
nchar(colnames(gbm_tcga_counts)[1])
nchar(colnames(gbm_tcga_counts))
gbm_tcga_counts
```
```{r}
#https://stackoverflow.com/questions/7963898/extracting-the-last-n-characters-from-a-string-in-r
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}
```

```{r}
ids<- substrRight(colnames(gbm_tcga_counts), 36)
ids<- str_replace_all(ids, "[[:punct:]]", "-")
```

```{r}
colnames(gbm_tcga_counts) <- ids
gbm_tcga_counts_order <- gbm_tcga_counts[,order(ids)]
gbm_tcga_metadata_order<- gbm_tcga_metadata[order(gbm_tcga_metadata$external_id), ]
```

```{r}
identical(colnames(gbm_tcga_counts_order), gbm_tcga_metadata_order$external_id)
```

```{r}
colnames(gbm_tcga_counts_order)<- gbm_tcga_metadata_order$tcga_barcode
```


tcgabiolinks download done by Avery Williams 
```{r}
load("/home/rstudio/data/TCGA_GBM_GeneExpression_TCGABIOlinks.rda")
```

```{r}
metadata_gbm<- colData(data)
```

```{r}
table(metadata_gbm$paper_IDH.status)
```

```{r}
#all primary tumors 
 metadata_gbm$sample_type[ metadata_gbm$paper_IDH.status == "WT"]
```

141 wildtype GBM. mutant IDH are no longer GBM 
```{r}
IDS <- metadata_gbm$barcode[ metadata_gbm$paper_IDH.status == "WT"]
IDS<- IDS[!is.na(IDS)]
#inculde the five nromal samples 
nt<- metadata_gbm$barcode[ metadata_gbm$sample_type == "Solid Tissue Normal"]
IDS<- c(IDS, nt)
```



```{r}
recount3_count_GBM<- gbm_tcga_counts_order[,colnames(gbm_tcga_counts_order) %in% IDS]
```

Download GTEx Brain dataset: make raw count and metadata table

```{r}
colnames(brain_gtex_counts)[1:5]
brain_gtex_metadata$external_id[1:5]
```


```{r}
ids<- str_replace_all(colnames(brain_gtex_counts), "[[:punct:]]", "-") 
meta_data_ids<- str_replace_all(brain_gtex_metadata$external_id, "[[:punct:]]", "-")
```

```{r}
identical(ids[order(ids)], meta_data_ids[order(meta_data_ids)])
```

```{r}
brain_gtex_counts_order <- brain_gtex_counts[,order(ids)]
brain_gtex_metadata_order<- brain_gtex_metadata[order(meta_data_ids), ]
```

```{r}
colnames(brain_gtex_counts_order )<- brain_gtex_metadata_order$external_id
```

```{r}
recount3_count_brain <- as.data.frame(brain_gtex_counts_order)
```


### Analysis

#pca analysis of both datasets 

TCGA GBM PCA analysis
```{r}
library(DESeq2)
vst_table <- vst(as.matrix(recount3_count_GBM))
vst_table_df <- t(vst_table)
pca.tumor <- prcomp(vst_table_df)
```

```{r}
summary(pca.tumor)
```

```{r}
recount_metadata<- gbm_tcga_metadata_order
#normal_ids<- rownames(recount_metadata)[recount_metadata$tcga_barcode %in% nt]

tumor_norm <- ifelse(colnames(recount3_count_GBM) %in% nt, "black", "red")
plot(pca.tumor$x[, 1], pca.tumor$x[, 2], pch = 20, col = tumor_norm  , main = "PCA of GBM", xlab = "PC1 (15.01%)", ylab = "PC2 (8.20%)", cex.axis = "1.5", cex.lab = "1.5")
legend("topleft", legend = c("tumor", "normal"), pch = 21, pt.bg = c("red", "black"), col = "black")
```

```{r}
sex<- metadata_gbm$barcode[metadata_gbm$gender == "female"]
sex<- sex[!is.na(sex)]
#normal_ids<- rownames(recount_metadata)[recount_metadata$tcga.tcga_barcode %in% sex]

tumor_norm <- ifelse(colnames(recount3_count_GBM) %in% sex, "red", "black")
plot(pca.tumor$x[, 1], pca.tumor$x[, 2], pch = 20, col = tumor_norm  , main = "PCA of GBM", xlab = "PC1 (15.06%)", ylab = "PC2 (8.22%)", cex.axis = "1.5", cex.lab = "1.5")
legend("topleft", legend = c("female", "male/normal"), pch = 21, pt.bg = c("red", "black"), col = "black")
```
no large covariates to account for in analysis

GTEx Brain PCA analysis
```{r}
library(DESeq2)
vst_table <- vst(as.matrix(recount3_count_brain))
vst_table_df <- t(vst_table)
pca.brain <- prcomp(vst_table_df)
```

```{r}
eigs <- pca.brain$sdev^2
summ <- rbind(
  SD = sqrt(eigs),
  Proportion = eigs/sum(eigs),
  Cumulative = cumsum(eigs)/sum(eigs))
```
```{r}
summ[,1:10]
```

```{r}
plot(pca.brain$x[,1], pca.brain$x[,2])
```

```{r}
group_samples<- names(pca.brain$x[,1])[pca.brain$x[,1] >40]
```

```{r}
metadata_gtex<- brain_gtex_metadata_order

metadata_gtex_pca1 <- metadata_gtex[metadata_gtex$external_id %in% group_samples,]
metadata_gtex_pca1.2 <- metadata_gtex[! metadata_gtex$external_id  %in% group_samples,]
```

```{r}
table(metadata_gtex_pca1$SMTSD)
```

```{r}
table(metadata_gtex_pca1.2$SMTSD)
```

```{r}
table(metadata_gtex_pca1$SEX)
```

```{r}
table(metadata_gtex_pca1.2$SEX)
```

note that brain structures are a major influence in the PCA, but I can not control this in DESeq2 or limma models because I don't have this information in TCGA GBM samples

only covariate at this point is GTex vs TCGA databases


#GBM vs normal DESeq2

create metadata for DESeq2 - database, tumor/normal, subtype, sex
```{r}
#order the tcga data to match recount3 order
rownames(recount_metadata)<- recount_metadata$tcga_barcode
recount3_ids <- colnames(recount3_count_GBM)
recount_metadata_sub <- recount_metadata[rownames(recount_metadata) %in% recount3_ids,]
identical(rownames(recount_metadata_sub),recount3_ids)
ids <- recount_metadata_sub$tcga_barcode

nam <- 
status<- c()
subtype<- c()
sex<- c()
for (i in 1:length(ids)){
  
nam[i]<- ids[i]
#create tumor/normal col
status[i] <-metadata_gbm$sample_type[metadata_gbm$barcode== ids[i]]
#create subtype col
subtype[i] <-as.character(metadata_gbm$paper_Transcriptome.Subtype[metadata_gbm$barcode == ids[i]])
subtype[i]<- ifelse(is.na(subtype[i]), "NA", subtype[i])
#create sex col 
sex[i]<- metadata_gbm$gender[metadata_gbm$barcode == ids[i]]
}
#create database col
database <- rep("TCGA", 146)
#create metadata 
tcga_gbm_metadata<- as.data.frame(cbind(nam, database, status, subtype, sex))
rownames(tcga_gbm_metadata)<- recount3_ids
tcga_gbm_metadata[1:5,]
```

Look at how many of each
```{r}
table(tcga_gbm_metadata$status)
table(tcga_gbm_metadata$subtype)
table(tcga_gbm_metadata$sex)
```

create metadata for GTEx
```{r}
#order the metadata to match recount3 order
rownames(metadata_gtex)<- metadata_gtex$external_id
#create sex col 
identical (rownames(metadata_gtex), colnames( recount3_count_brain))
nam<- metadata_gtex$external_id
sex<- ifelse(metadata_gtex$SEX == 1, "male", "female")
#create database col
database<- rep ("GTEx",2931 )
#create normal col
status<- rep ("Solid Tissue Normal", 2931)
#create subtype col
subtype<- rep("NA", 2931)
#create metadata 
gtex_metadata<- as.data.frame(cbind(nam, database, status, subtype, sex))
rownames(gtex_metadata)<- rownames(metadata_gtex)
#combine count and metadata
gtex_metadata[1:5,]
```


```{r}
all_metadata <- rbind(tcga_gbm_metadata, gtex_metadata)
rownames(all_metadata)<- c(rownames(tcga_gbm_metadata), rownames(gtex_metadata))
count_table<- cbind(recount3_count_GBM, recount3_count_brain)
```

```{r}
identical(rownames(all_metadata), colnames(count_table))
```

```{r eval = FALSE}
saveRDS(count_table, "/home/rstudio/data/recount3_GBM_GTEx_BRAIN_raw_counts.rds")
saveRDS(all_metadata, "/home/rstudio/data/recount3_GBM_GTEx_BRAIN_metadata.rds")
```


Create Deseq2 object
```{r}
dds_gbm_normal <- DESeqDataSetFromMatrix(countData = count_table,
                              colData = all_metadata,
                              design= ~ database + status)
```

Run DESeq2 for GBM tumor vs normal
```{r eval = FALSE}
deseq2_gbm_normal <- deseq2_function(dds_gbm_normal, "~/output/deseq2_gbm/220927_deseq2_gbm_normal_gtx_res.csv", "status", "Primary Tumor", "Solid Tissue Normal")
```
estimating size factors

  Note: levels of factors in the design contain characters other than

  letters, numbers, '_' and '.'. It is recommended (but not required) to use

  only letters, numbers, and delimiters '_' or '.', as these are safe characters

  for column names in R. [This is a message, not a warning or an error]

estimating dispersions

gene-wise dispersion estimates

mean-dispersion relationship

  Note: levels of factors in the design contain characters other than

  letters, numbers, '_' and '.'. It is recommended (but not required) to use

  only letters, numbers, and delimiters '_' or '.', as these are safe characters

  for column names in R. [This is a message, not a warning or an error]

final dispersion estimates

  Note: levels of factors in the design contain characters other than

  letters, numbers, '_' and '.'. It is recommended (but not required) to use

  only letters, numbers, and delimiters '_' or '.', as these are safe characters

  for column names in R. [This is a message, not a warning or an error]
fitting model and testing

  Note: levels of factors in the design contain characters other than

  letters, numbers, '_' and '.'. It is recommended (but not required) to use

  only letters, numbers, and delimiters '_' or '.', as these are safe characters

  for column names in R. [This is a message, not a warning or an error]

-- replacing outliers and refitting for 15181 genes

-- DESeq argument 'minReplicatesForReplace' = 7 

-- original counts are preserved in counts(dds)

estimating dispersions

  Note: levels of factors in the design contain characters other than

  letters, numbers, '_' and '.'. It is recommended (but not required) to use

  only letters, numbers, and delimiters '_' or '.', as these are safe characters

  for column names in R. [This is a message, not a warning or an error]

fitting model and testing

  Note: levels of factors in the design contain characters other than

  letters, numbers, '_' and '.'. It is recommended (but not required) to use

  only letters, numbers, and delimiters '_' or '.', as these are safe characters

  for column names in R. [This is a message, not a warning or an error]

[1] "Intercept"                                   "database_TCGA_vs_GTEx"                      

[3] "status_Primary.Tumor_vs_Solid.Tissue.Normal"

using 'apeglm' for LFC shrinkage. If used in published research, please cite:

    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for

    sequence count data: removing the noise and preserving large differences.

    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895

```{r eval = FALSE}
deseq2_gbm_normal
```



#Transfer learning GBM vs Normal
normalize count data by tpm

get gene lengths

check to make sure things line up 
```{r}
identical(feature_info$gene_id, rownames(count_table))
```

```{r}
tpm_table<- Counts_to_tpm(count_table, feature_info$bp_length)
```

```{r}
rownames(tpm_table)<- feature_info$gene_name
```


```{r eval=FALSE}
saveRDS(tpm_table, "~/output/recount3_gbm_gtex_tpm_df.rds")
```



differential latent variable analysis

bring in PLIER object
```{r eval= FALSE}
recount3_plier <- readRDS("~/data/recount3/recount3_plier_wo.rds")
```

```{r eval= FALSE}
#There is a issue with handling genes with zero expression. Best thing to do is to remove the genes
test <- GetOrderedRowNorm(as.matrix(tpm_table),recount3_plier )
zeros<- rownames(test)[!complete.cases(test)]
tpm_table_v2<- tpm_table[!rownames(tpm_table) %in% zeros,]
```

create the b.matrix 
```{r eval= FALSE}
GBM.GTEx.b.matrix <- GetNewDataB(as.matrix(tpm_table_v2), recount3_plier)
```

```{r eval=FALSE}
saveRDS(GBM.GTEx.b.matrix, file= "~/output/TF_L_GBM/220804_GBM_GTEX_B_MATRIX.rds")
```

Run limma for differential latent variable analyis 
LVTestWrapperJLF <- function(b.matrix,
                          sample.info.df,
                          phenotype.col,
                          blocking,
                          file.lead,
                          plot.dir = "plots",
                          results.dir = "results",
                          use.bonferroni = FALSE,
                          significant.only = FALSE,
                          sig.threshold = 0.05)
```{r eval= FALSE}
#adjust metadata for the function
sampleinfo<- cbind(colnames(GBM.GTEx.b.matrix), as.character(all_metadata$status), as.character(all_metadata$database))
colnames(sampleinfo) <- c("Sample", "Tumor","database")
sampleinfo<- as.data.frame(sampleinfo)
database<- all_metadata$database
names(database)<- colnames(GBM.GTEx.b.matrix)

LV_test_TCGA_BH <- LVTestWrapperJLF(b.matrix=GBM.GTEx.b.matrix , sample.info.df= sampleinfo, phenotype.col= "Tumor", blocking= database, file.lead= "220804_LV_TCGA_GTEX_T_vs_NT", plot.dir= "~/output/TF_L_GBM/", results.dir= "~/output/TF_L_GBM/", use.bonferroni= FALSE, sig.threshold = 0.05)
```

```{r eval=FALSE}
saveRDS(LV_test_TCGA_BH , file= "~/output/TF_L_GBM/220804_GBM_GTEX_limma_res.rds")
```


are genes and heatmaps and pathway analysis

```{r eval= FALSE}
sig_LVs <- LV_test_TCGA_BH$limma[LV_test_TCGA_BH$limma$adj.P.Val < 0.05,]
```


# END
    Location of final scripts:
    scripts
    
    Location of data produced:
    output 
    
    Dates when operations were done:
    220112; 220802 updated
    
## Versions
```{r}
sessionInfo()
```



## Code used to donwload TCGA biolinks data
# TCGAbiolinks Downloads
#via Avery Williams
library(SummarizedExperiment)
library(TCGAbiolinks)
# GDCquery: searches the genomic data commons for the given project and category, filtering results as needed and returning an object with a summary table
query.gbmexp <- GDCquery(project = "TCGA-GBM",
                         data.category = "Transcriptome Profiling",
                         access = "open",
                         legacy = FALSE,
                         experimental.strategy = "RNA-Seq",
                         workflow.type = "HTSeq - Counts")
# Downloading data to local computer
GDCdownload(query = query.gbmexp, method = "api")
# Bringing data into environment
TCGAdata <- GDCprepare(query = query.gbmexp, save = TRUE, 
                   save.filename = "TCGA_GBM_GeneExpression.rda",
                   summarizedExperiment = TRUE)
