---
title: "220405_limma_GBM"
author: "Jennifer Fisher"
date: "4/5/2022"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# START 
    Goal/Purpose of operations: 
    Apply limma differential gene expression analysis to GBM and GTEx controls to identify genes that differential expressed between tumor and control gene expression profiles. The differential expressed gene were used to signature reversion. 
    
    Finished psedocode on: 
    220405
    
    System which operations were done on: 
    my laptop
    
    GitHub Repo:
    Transfer_Learning_R03
    
    Docker/Singularity: 
    rstudio_tf_dr_v3
    
    Directory of operations:
    "/home/rstudio"
    
    Scripts being edited for operations:
    using some modified scripts from MultiPLIER paper
    
    Data being used: 
    TCGA GBM data and GTEx Brain samples
    
    Papers and tools: 
    limma 
    MultiPLIER

# STEPS
### Set working directory 
### load in data
```{r}
library(recount3)
library(SummarizedExperiment)
library(dplyr)
```
load in function scripts 
```{r}
source("/home/rstudio/script/functions_cancer_signature_reversion_JLF.R")
source("/home/rstudio/script/plier_util.R")
source("/home/rstudio/script/test_LV_differences.R")
```

```{r}
count_table<- readRDS( "/home/rstudio/data/recount3_GBM_GTEx_BRAIN_raw_counts.rds")
all_metadata<- readRDS("/home/rstudio/data/recount3_GBM_GTEx_BRAIN_metadata.rds")
```

normalize count data by tpm

get the gene lengths from recount3 
```{r}
human_projects<- available_projects()
recount3_rse_BRIAN <- create_rse(human_projects[(human_projects$project == "BRAIN"),])
gene_info <- as.data.frame(recount3_rse_BRIAN@rowRanges)
```
check to make sure things line up 
```{r}
identical(gene_info$gene_id, rownames(count_table))
```

```{r}
tpm_table<- Counts_to_tpm(count_table, gene_info$bp_length)
```

```{r}
rownames(tpm_table)<- gene_info$gene_name
```

```{r}
tpm_df <- as.data.frame(tpm_table)
```

### Analysis

## limma analysis 

```{r}
rm(count_table, recount3_rse_BRIAN)
gc()
```

transfrom TPM for limma differential gene expression
```{r}
TPM <- log2(as.matrix(tpm_df) + 1)
```

                          
```{r eval= FALSE}
#adjust metadata for the function
sampleinfo<- cbind(colnames(tpm_df), as.character(all_metadata$status), as.character(all_metadata$database))
colnames(sampleinfo) <- c("Sample", "Tumor","database")
sampleinfo<- as.data.frame(sampleinfo)
database<- all_metadata$database
names(database)<- colnames(tpm_df)

limma_gene_TCGA_BH <- LVTestWrapperJLF_v2(b.matrix=TPM , sample.info.df= sampleinfo, phenotype.col= "Tumor", blocking= database, file.lead= "220421_limma_TCGA_GTEX_T_vs_NT", plot.dir= "~/output/limma_gbm/", results.dir= "~/output/limma_gbm/", use.bonferroni= FALSE, sig.threshold = 0.05)
```


```{r eval=FALSE}
saveRDS(limma_gene_TCGA_BH , file= "~/output/limma_gbm/220421_GBM_GTEX_gene_limma_res.rds")
```




## signaturesearch 
get the limma results
```{r}
GBM_GTEX_gene_limma_res <- readRDS("~/output/limma_gbm/220421_GBM_GTEX_gene_limma_res.rds")
```

get LINCS 2020 Level 5 data
```{r}
library(signatureSearch)
library(HDF5Array)
se <- SummarizedExperiment(HDF5Array("~/data/lincs_2020.h5", name="assay"))
```

```{r}
rownames(se) <- HDF5Array("~/data/lincs_2020.h5", name="rownames")
colnames(se) <- HDF5Array("~/data/lincs_2020.h5", name="colnames")
db_path <- "~/data/lincs_2020.h5"
```

```{r}
limma_results<- GBM_GTEX_gene_limma_res$limma
```

get the top up and down regulated genes; use the top ~100 from the up and down regulated genes
```{r}
limma_results<- limma_results[complete.cases(limma_results),]
limma_results_sig <- limma_results[limma_results$adj.P.Val < 0.05 & abs(limma_results$logFC) > 2,]
```

```{r}
upset<-limma_results_sig$LV[order(-limma_results_sig$logFC)]
downset<- limma_results_sig$LV[order(limma_results_sig$logFC)]
upset<- upset[1:200]
downset<- downset[1:200]
```
convert gene symbols to LINCS genes
```{r}
downset_v2 <- match_to_LINCS_genes(upset)
upset_v2 <- match_to_LINCS_genes(downset)
```

```{r}
downset_v2<- downset_v2[complete.cases(downset_v2)]
upset_v2<- upset_v2[complete.cases(upset_v2)]
```

```{r}
names(upset_v2)<- NULL
names(downset_v2)<- NULL
```

```{r}
limma_gbm_list<- list(up= upset_v2, down= downset_v2)
```
save the input gene signature. 
```{r}
saveRDS(limma_gbm_list,"~/output/limma_gbm/SR_gene_list_gbm_limma.rds")
```

run signature reversion for limma GBM vs Control
```{r}
set.seed(101)
qsig_lincs_limma_res <- qSig(query=list(upset=as.character(upset_v2), downset=as.character(downset_v2)), 
                   gess_method="LINCS", refdb=db_path)
```

```{r}
set.seed(101)
lincs_limma_gbm <- gess_lincs(qsig_lincs_limma_res, sortby="NCS", tau=TRUE, workers=1)
#result(lincs)
```

```{r}
lincs_limma_gbm_res <- result(lincs_limma_gbm)
```
```{r}
write.csv(lincs_limma_gbm_res , file= "~/output/limma_gbm/220421_SR_LINCS_GBM_LIMMA_RES.csv" )
```


tau calc 

```{r}
lincs_limma_gbm_res_GI1 <- lincs_limma_gbm_res[lincs_limma_gbm_res$cell == "GI1",]
```

```{r}
GI1_taudf <- readRDS("~/data/GI1_taudf.rds")
```

```{r}
tau_test<- tau_scores(lincs_limma_gbm_res_GI1, GI1_taudf)
```

```{r}
lincs_limma_gbm_res_GI1$jlf_tau<-tau_test
```

```{r}
limma_sig_drugs <- lincs_limma_gbm_res_GI1[lincs_limma_gbm_res_GI1$WTCS_Pval <0.05 & lincs_limma_gbm_res_GI1$NCS <0 & as.numeric(lincs_limma_gbm_res_GI1$jlf_tau) < -80, ] 
```

drug analysis

```{r}
#FDA_APPROVAL_CHECK
lincs_limma_gbm_res_GI1$fda_database<- FDA_APPROVAL_CHECK(lincs_limma_gbm_res_GI1$pert)
```

```{r}
limma_sig_drugs_fda <-limma_sig_drugs[ limma_sig_drugs$fda_database == TRUE,]
```
7/16 drugs are in clinical trial for GBM 

```{r}
hyperG_k_res_limma <- dsea_hyperG(drugs =limma_sig_drugs_fda$pert , type = "KEGG", 
                            pvalueCutoff = 1, qvalueCutoff = 1, 
                            minGSSize = 10, maxGSSize = 2000)
hyperG_k_res_limma_v2<- result(hyperG_k_res_limma)
```

```{r}
#false discovery rate
hyperG_k_res_limma_v2$logqvalue<- -log(hyperG_k_res_limma_v2$qvalue)
```

```{r}
hyperG_k_res_limma_v2$Description<- factor(hyperG_k_res_limma_v2$Description, levels= hyperG_k_res_limma_v2$Description)
```

```{r}
gbm.2.2<-ggplot(data=hyperG_k_res_limma_v2[1:30,] , aes(x=Description, y=Count, fill=logqvalue )) +
  geom_bar(stat="identity") 
gbm.2.2 + coord_flip(ylim = c(0,8)) + ylab("Number of Drugs") + xlab("KEGG Pathways") +labs(fill = "-log(q-value)")
ggsave("~/output/limma_gbm/220421_ggplot_limma_tau_gbm_approved_candidates_pathways.png")
```

check that all the drug can result in pathways== yes
```{r}
limma_sig_drugs_fda$pert <- factor(limma_sig_drugs_fda$pert, levels =limma_sig_drugs_fda$pert[order(-limma_sig_drugs_fda$NCS)] )
p<-ggplot(data=limma_sig_drugs_fda , aes(x=pert, y=NCS, fill=jlf_tau)) +
  geom_bar(stat="identity")+ scale_fill_viridis(direction=-1)
p + coord_flip()  + ylab("Normalized Connectivity Score") + xlab("LINCS Small Molecules") +labs(fill = "Tau [-100,100]")
ggsave("~/output/limma_gbm/2200421_ggplot_limma_tau_gbm_approved_candidates.png")
```

    
# END
    Location of final scripts:
    script
    
    Location of data produced:
    ~/output/limma_gbm
    
    Dates when operations were done:
    2200421; updated 220802
    
## Versions
```{r}
sessionInfo()
```

