---
title: "220527_plier_LIHC"
author: "Jennifer Fisher"
date: "5/27/2022"
output:
  html_document:
    toc: true
    theme: united
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# START 
    Goal/Purpose of operations: 
    Conducting Demension reduction for Transfer learning with PLIER
    
    Finished psedocode on: Date
    220104
    
    System which operations were done on:
    my laptop(on cheaha)
    
    GitHub Repo:
    Transfer_Learning_R03
    
    Docker/Singularity: 
    rstudio_tf_dr_v3
    
    Directory of operations: 
    /data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03
    
    Scripts being edited for operations:
    NA
    
    Data being used:
    recount3- tpm normalized
    recount3 is an online resource consisting of RNA-seq gene, exon, and exon-exon junction counts as well as coverage bigWig files for 8,679 and 10,088 different studies for human and mouse respectively. It is the third generation of the ReCount project and part of recount.bio.
    http://rna.recount.bio/
    ./data/recount3
    
    Papers and tools: 
    recount3- see details above 
    PLIER and MultiPLIER wrapper functions for PLIER

# STEPS
### Set working directory 
/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03

### load in data
```{r}
recount3_group1_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group1_tpm.rds")
recount3_group2_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group2_tpm.rds")
recount3_group3_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group3_tpm.rds")
recount3_group4_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group4_tpm.rds")
recount3_group5_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group5_tpm.rds")
recount3_group6_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group6_tpm.rds")
recount3_group7_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group7_tpm.rds")
recount3_group8_tpm <- readRDS("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_group8_tpm.rds")
```

```{r}
recount3_tpm <- cbind(recount3_group1_tpm, recount3_group2_tpm, recount3_group3_tpm, recount3_group4_tpm, recount3_group5_tpm, recount3_group6_tpm, recount3_group7_tpm, recount3_group8_tpm)
```

```{r}
rm(recount3_group1_tpm, recount3_group2_tpm, recount3_group3_tpm, recount3_group4_tpm, recount3_group5_tpm, recount3_group6_tpm, recount3_group7_tpm, recount3_group8_tpm)
gc()
```

remove Nas

```{r}
dim(recount3_tpm[1:4,is.na(recount3_tpm[1,])])
```


```{r}
recount3_tpm_v2 <- recount3_tpm[, !is.na(recount3_tpm[1,])]
```


### Analysis
```{r}
library(PLIER)
```

```{r}
source("/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/script/plier_util.R")
```

```{r}
library(recount3)
```

```{r}
human_samples<- available_samples()
```
```{r}
human_projects<- available_projects()
```
```{r}
#SRP118922
test<- create_rse(human_projects[(human_projects$project == "SRP118922"),])
```
```{r}
gene_info <- test@rowRanges
rm(test)
```


```{r}
human_samples_not_lihc<- human_samples$external_id[! human_samples$project == "LIHC" | human_samples$project  == "LIVER"]
human_samples_not_lihc_v2 <-gsub("-", ".", human_samples_not_lihc)
```

```{r}
library(dplyr)
recount3_wo_lihc_tpm <- recount3_tpm_v2[, colnames(recount3_tpm_v2) %in% human_samples_not_lihc_v2 ]

```

```{r}
saveRDS(recount3_wo_lihc_tpm, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_wo_lihc_tpm.rds")
```




```{r}
rm(human_samples, human_samples_not_lihc, recount3_tpm, recount3_tpm_v2)
gc()
```


```{r}
recount3_wo_lihc_tpm<- as.matrix(recount3_wo_lihc_tpm)
rownames(recount3_wo_lihc_tpm) <- gene_info$gene_name
```


```{r}
PLIERNewData_JLF <- function(exprs.mat, seed = 12345) {
  # A wrapper function for applying PLIER to a data set. We use the following
  # genesets that come with PLIER: bloodCellMarkersIRISDMAP, svmMarkers, 
  # and canonicalPathways. We set the k parameter for the PLIER model by
  # identifying the number of "significant PCs" with PLIER::num.pc and then 
  # using sig PCs * 0.3. This is consistent with recommendations from the 
  # PLIER authors.
  # 
  # Args:
  #   exprs.mat: a gene expression matrix, rows are genes, columns are samples
  #   seed: an integer to be supplied to set.seed() for reproducibility 
  #         purposes, default is 12345
  #         
  # Returns:
  #   plier.res: output from PLIER::PLIER()
  #
  require(PLIER)
  
  set.seed(seed)
  
  # load PLIER pathway and cell type data
  data(bloodCellMarkersIRISDMAP)
  data(svmMarkers)
  data(canonicalPathways)
  
  print("combine the pathway data from PLIER")
  all.paths <- PLIER::combinePaths(bloodCellMarkersIRISDMAP, svmMarkers, 
                                   canonicalPathways)
  
  # what genes are common to the pathway data and the expression matrix
  cm.genes <- PLIER::commonRows(all.paths, exprs.mat)
  
  print(" row normalize")
  exprs.norm <- PLIER::rowNorm(exprs.mat)
  
  # what should we set the minimum k parameter to in PLIER? estimate the number 
  # of PC for the SVD decomposition 
  print("num.pc")
  mtx <- exprs.norm[cm.genes, ]
  mtx<- mtx[complete.cases(mtx),]
  set.k <- PLIER::num.pc(mtx )
  
  # PLIER main function + return results
  print("PLIER main function")
  plier.res <- PLIER::PLIER(mtx, all.paths[cm.genes, ], scale= F,
                           k = round((set.k + set.k * 0.3), 0), trace = TRUE)
  
  return(plier.res)
  
}
```


```{r}
recount3_plier_lihc <- PLIERNewData_JLF(recount3_wo_lihc_tpm)
saveRDS(recount3_plier_lihc, "/data/project/lasseigne_lab/JLF_scratch/Transfer_Learning_R03/data/recount3/recount3_plier_lihc.rds")
```


```{r}
PLIERNewData_JLF_step1<- function(exprs.mat, seed = 12345) {
  # A wrapper function for applying PLIER to a data set. We use the following
  # genesets that come with PLIER: bloodCellMarkersIRISDMAP, svmMarkers, 
  # and canonicalPathways. We set the k parameter for the PLIER model by
  # identifying the number of "significant PCs" with PLIER::num.pc and then 
  # using sig PCs * 0.3. This is consistent with recommendations from the 
  # PLIER authors.
  # 
  # Args:
  #   exprs.mat: a gene expression matrix, rows are genes, columns are samples
  #   seed: an integer to be supplied to set.seed() for reproducibility 
  #         purposes, default is 12345
  #         
  # Returns:
  #   plier.res: output from PLIER::PLIER()
  #
  require(PLIER)
  
  set.seed(seed)
  
  # load PLIER pathway and cell type data
  data(bloodCellMarkersIRISDMAP)
  data(svmMarkers)
  data(canonicalPathways)
  
  print("combine the pathway data from PLIER")
  all.paths <- PLIER::combinePaths(bloodCellMarkersIRISDMAP, svmMarkers, 
                                   canonicalPathways)
  
  # what genes are common to the pathway data and the expression matrix
  cm.genes <- PLIER::commonRows(all.paths, exprs.mat)
  
  print(" row normalize")
  exprs.norm <- PLIER::rowNorm(exprs.mat)
  
  # what should we set the minimum k parameter to in PLIER? estimate the number 
  # of PC for the SVD decomposition 
  print("num.pc")
  mtx <- exprs.norm[cm.genes, ]
  mtx<- mtx[complete.cases(mtx),]
  set.k <- PLIER::num.pc(mtx )
  my_list <- list("mtx" = mtx, "all.paths" = all.paths, "cm.genes" = cm.genes, "set.k"=set.k)
  return(my_list)
}

PLIERNewData_JLF_step2<- function(mtx, all.paths, cm.genes, set.k){
  # PLIER main function + return results
  print("PLIER main function")
  plier.res <- PLIER::PLIER(mtx, all.paths[cm.genes, ], scale= F,
                           k = round((set.k + set.k * 0.3), 0), trace = TRUE)
  
  return(plier.res)
  
}
```

